autoretrieve/engine.go

```diff
@@ -0,0 +1,524 @@
+	if e.publisher != nil {
+		log := log.With("adCid", c)
+		log.Info("Publishing advertisement in pubsub channel")
+		err = e.publisher.UpdateRoot(ctx, c)
+		if err != nil {
+			log.Errorw("Failed to announce advertisement on pubsub channel ", "err", err)
+			return cid.Undef, err
+		}
+	}
+	return c, nil
+}
+
+// PublishLatest re-publishes the latest existing advertisement to pubsub.
+func (e *AutoretrieveEngine) PublishLatest(ctx context.Context) error {
+	// Skip announcing the latest advertisement CID if there is no publisher.
+	if e.publisher == nil {
+		log.Infow("Skipped announcing the latest: remote announcements are disabled.")
+		return nil
+	}
+
+	adCid, err := e.getLatestAdCid(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get latest advertisement cid: %w", err)
+	}
+
+	if adCid == cid.Undef {
+		log.Info("Skipped announcing the latest: no previously published advertisements.")
+		return nil
+	}
+	log.Infow("Republishing latest advertisement", "cid", adCid)
+
+	return e.publisher.UpdateRoot(ctx, adCid)
+}
+
+// RegisterMultihashLister registers a provider.MultihashLister that is used to look up the
+// list of multihashes associated to a context ID. At least one such registration
+// must be registered before calls to AutoretrieveEngine.NotifyPut and AutoretrieveEngine.NotifyRemove.
+//
+// Note that successive calls to this function will replace the previous registration.
+// Only a single registration is supported.
+//
+// See: provider.Interface
+func (e *AutoretrieveEngine) RegisterMultihashLister(mhl provider.MultihashLister) {
+	log.Debugf("Registering multihash lister in engine")
+	e.cblk.Lock()
+	defer e.cblk.Unlock()
+	e.mhLister = mhl
+}
+
+// NotifyPut publishes an advertisement that signals the list of multihashes
+// associated to the given contextID is available by this provider with the
+// given metadata. A provider.MultihashLister is required, and is used to look up the
+// list of multihashes associated to a context ID.
+//
+// Note that prior to calling this function a provider.MultihashLister must be registered.
+//
+// See: AutoretrieveEngine.RegisterMultihashLister, AutoretrieveEngine.Publish.
+func (e *AutoretrieveEngine) NotifyPut(ctx context.Context, contextID []byte, providerID string, addresses []string, md metadata.Metadata) (cid.Cid, error) {
+	// The multihash lister must have been registered for the linkSystem to know how to
+	// go from contextID to list of CIDs.
+	return e.publishAdvForIndex(ctx, contextID, providerID, addresses, md, false)
+}
+
+// NotifyRemove publishes an advertisement that signals the list of multihashes associated to the given
+// contextID is no longer available by this provider.
+//
+// Note that prior to calling this function a provider.MultihashLister must be registered.
+//
+// See: AutoretrieveEngine.RegisterMultihashLister, AutoretrieveEngine.Publish.
+func (e *AutoretrieveEngine) NotifyRemove(ctx context.Context, contextID []byte) (cid.Cid, error) {
```

Comment:
I want to assume we will use this for when contents are offloaded, right?

Reply:
Maybe not even then. This is a question I had for @whyrusleeping: even when content is offloaded, we're still able to retrieve it for users right? So theoretically we would never need to remove anything from the advertisements.